### PR TITLE
Check that target features required by LLVM intrinsics are enabled

### PR DESCRIPTION
**Pull request by eduardosm**
Fixes https://github.com/rust-lang/miri/issues/3178

### CODE REVIEW

**Comment by eduardosm** [comment]
I'm not sure what's going on. `llvm.arm.hint` is called through `std::arch::arm::__yield`, which requires `target_feature = "v6"`.

```rust
#[cfg(any(target_feature = "v6", target_arch = "aarch64", doc))]
#[inline(always)]
pub unsafe fn __yield() {
    hint(HINT_YIELD);
}

...

extern "unadjusted" {
    #[cfg_attr(target_arch = "aarch64", link_name = "llvm.aarch64.hint")]
    #[cfg_attr(target_arch = "arm", link_name = "llvm.arm.hint")]
    fn hint(_: i32);
}
```

File: src/helpers.rs [hunk 1]
```diff
@@ -1063,6 +1063,24 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
...
+    /// required by `intrinsic`.
+    fn expect_target_feature_for_intrinsic(
+        &self,
+        intrinsic: Symbol,
+        target_feature: Symbol,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_ref();
+        if !this.tcx.sess.unstable_target_features.contains(&target_feature) {
+            throw_ub_format!(
+                "attempted to call intrinsic `{intrinsic}` that requires missing {target_feature} target feature"
```
Comment (RalfJung) [inline src/helpers.rs:1079]:
```suggestion
                "attempted to call intrinsic `{intrinsic}` that requires missing target feature `{target_feature}`"
```

File: src/shims/foreign_items.rs [hunk 2]
```diff
@@ -1008,9 +1008,11 @@ trait EvalContextExtPriv<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
             "llvm.arm.hint" if this.tcx.sess.target.arch == "arm" => {
                 let [arg] = this.check_shim(abi, Abi::Unadjusted, link_name, args)?;
                 let arg = this.read_scalar(arg)?.to_i32()?;
+                // Note that different arguments might have different target feature requirements.
                 match arg {
                     // YIELD
                     1 => {
+                        this.expect_target_feature_for_intrinsic(link_name, Symbol::intern("v6"))?;
```
Comment (RalfJung) [inline src/shims/foreign_items.rs:1015]: Looks like `expect_target_feature_for_intrinsic` should take `&str` as its second argument?

**Review by RalfJung — COMMENTED** [review]
Implementation looks good, I just hope we got them all. :)

Could you add a test to ensure the UB indeed does get reported properly?

File: src/shims/foreign_items.rs (hunk 2 above)
Comment (eduardosm) [reply src/shims/foreign_items.rs:1015]: Yeah, since we only have `sse` and `avx` in `sym` anyways

**Comment by eduardosm** [comment]
Oh right, SSE4.1 is enabled by default on macOS

File: tests/fail/shims/intrinsic_target_feature.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+
+fn main() {
+    assert!(is_x86_feature_detected!("sse"));
+    assert!(!is_x86_feature_detected!("sse4.1"));
+
+    unsafe {
+        // Pass, since SSE is enabled
+        let _ = addss(_mm_setzero_ps(), _mm_setzero_ps());
```
Comment (RalfJung) [inline tests/fail/shims/intrinsic_target_feature.rs:27]: Are these must_use, or why the `let _`?

Comment (eduardosm) [reply tests/fail/shims/intrinsic_target_feature.rs:27]: Simply to make it explicit that we don't care about the returned value

Comment (RalfJung) [reply tests/fail/shims/intrinsic_target_feature.rs:27]: I find this more confusing than helpful, TBH. I'd prefer if you removed it.

Comment (eduardosm) [reply tests/fail/shims/intrinsic_target_feature.rs]: So be it

**Comment by RalfJung** [comment]
Thanks for the PR!
@bors r+

**Pull request merged**